### ISSUE TITLE
A bug in your clustering programming

### ISSUE THREAD

**Issue by justopit** [vana77/Bottom-up-Clustering-Person-Re-identification#11]
From the code, you  calculates distance between all features of images (including penalize with unbalanced number of groups) in advance when clustering. However, you ignore a problem, that in the process of clustering, a cluster may be combined many times, resulting in changes in the number of groups and distance would change. For example,  cluster 2 is first combined with cluster 3.  However, when cluster 2 is then combined with 4, the cluster number is changed.

**Comment by vana77** [comment]
Hi @justopit , I don't think there is a bug. In this case, cluster2, 3, 4 are merged into one cluster, and the number of cluster is changed to n-2.

**Comment by justopit** [comment]
@vana77 , I guess you misunderstood me, maybe for I didn't explain it well. I mean when cluster A is then combined with C(after combined with B). real distance between them is  min(D(A, B), D(A,C)) + lambda * (|A| + |B| + |C| ). However, because you calculate the distance matrix in advance, the distance between A and C becomes D(A,C) + lambda * (|A| + |C|).

**Comment by vana77** [comment]
Hi @justopit, yes, in traditional hierarchical clustering, in each iteration, only two clusters are merged, and then the distance is calculated again for the next iteration. In this paper, we use mp to control the number of clusters to be merged in one iteration, and we set mp to 0.05. When mp is set to 1/N, it will be the ideal situation. We did experiments to evaluate mp, and find that a smaller mp leads to better performance. However, we have to make a balance between the performance and time, which satisfies the accuracy.
